src/parsers/open-api3.js

```diff
@@ -3,38 +3,119 @@ const Validators = require('../validators/index'),
+    } else {
+        return new Validators.SimpleValidator(ajv.compile(requestDereferenceBody));
     }
-    const bodySchemaV3 = dereferenced.paths[currentPath][currentMethod].requestBody.content['application/json'].schema;
+}
+
+function buildRequestBodyValidation(dereferenced, referenced, currentPath, currentMethod, { ajvConfigBody, formats, keywords }) {
+    let requestDereferenceBody = getRequestSchema(dereferenced, currentPath, currentMethod);
+    let requestReferenceBody = getRequestSchema(referenced, currentPath, currentMethod);
+
+    if (!requestDereferenceBody || !requestReferenceBody) return;
+
     const defaultAjvOptions = {
         allErrors: true
     };
+
     const ajvOptions = Object.assign({}, defaultAjvOptions, ajvConfigBody);
     let ajv = new Ajv(ajvOptions);
 
     ajvUtils.addCustomKeyword(ajv, formats, keywords);
 
-    if (bodySchemaV3.discriminator) {
-        return buildV3Inheritance(dereferenced, originalSwagger, currentPath, currentMethod, ajv);
+    if (requestDereferenceBody.discriminator) {
+        let referencedSchemas = referenced.components.schemas;
+        let dereferencedSchemas = dereferenced.components.schemas;
+        let referenceName = requestReferenceBody['$ref'];
+
+        return buildV3Inheritance(referencedSchemas, dereferencedSchemas, currentPath, currentMethod, ajv, referenceName);
     } else {
-        return new Validators.SimpleValidator(ajv.compile(bodySchemaV3));
+        return new Validators.SimpleValidator(ajv.compile(requestDereferenceBody));
     }
 }
 
-function buildV3Inheritance(dereferencedDefinitions, swagger, currentPath, currentMethod, ajv) {
+function buildHeadersValidation(responses, { ajvConfigParams, formats, keywords, contentTypeValidation }, statusCode) {
+    let headers = responses[statusCode].headers;
+    if (!headers) return;
+
+    const defaultAjvOptions = {
+        allErrors: true,
+        coerceTypes: 'array'
+    };
+    const ajvOptions = Object.assign({}, defaultAjvOptions, ajvConfigParams);
+    let ajv = new Ajv(ajvOptions);
+
+    ajvUtils.addCustomKeyword(ajv, formats, keywords);
+
+    var ajvHeadersSchema = {
+        title: 'HTTP headers',
+        type: 'object',
+        properties: {},
+        required: [],
+        additionalProperties: true
+    };
+
+    Object.keys(headers).forEach(key => {
+        let headerObj = Object.assign({}, headers[key].schema);
+        const headerName = key.toLowerCase();
+        delete headerObj.name;
+        delete headerObj.required;
+        ajvHeadersSchema.properties[headerName] = headerObj;
+    });
+
+    ajvHeadersSchema.content = optionUtils.createContentTypeHeaders(contentTypeValidation, OAI3_RESPONSE_CONTENT_TYPE);
+
+    return new Validators.SimpleValidator(ajv.compile(ajvHeadersSchema));
+}
+
+function buildV3Inheritance(referencedSchemas, dereferencedSchemas, currentPath, currentMethod, ajv, referenceName) {
```

Comment:
why we need referencedSchemas and dereferencedSchemas?
oneOf can extract from both